### ISSUE TITLE
Support n-best lists in marian-scorer

### ISSUE THREAD

**Issue by emjotde** [marian-nmt/marian-dev#105]
Hi all,
I've added an option to the rescorer to summarize scores, so it can now calculate cross-entropy and perplexity for a whole test set and report it at the end. This is currently in the `nematus-dynamic` branch, which I will soon merge into master. 

I think the rescorer would become more useful if it could also do the following things:

* Read an n-best list and produce a new n-best list with an added new feature (feature name should be specifiable). The n-best list should be unchanged otherwise. 
* Produce an n-best list for plain text input (assuming constant specified N or N=1 in the default case), again with a given feature name.

**Comment by emjotde** [comment]
Maybe the `rescorer` should also just be renamed to `scorer`?

**Comment by emjotde** [comment]
And multi-gpu option, currently the rescorer only uses one GPU, there is no reason for that.

**Comment by snukky** [comment]
What about:
* `rescorer` -> `marian-scorer`
* `s2s` -> `marian-decoder` or `marian-translator`
* `server` -> `marian-server`

It's sometimes annoying to explain people who are new in the toolkit that s2s is a decoder. Also keeping a common prefix is reasonable when installing marian globally in the system.

**Comment by emjotde** [comment]
Makes sense.

**Comment by lkfo415579** [comment]
@emjotde thanks for this feature. But i don't see any document or usage about rescoring, would you provide your usage of rescoring to us? I tried to import n-best output file to the rescorer by using reversed language pair model, then it outputed some negative score file, i'm confused, did i import correctly?

**Comment by snukky** [comment]
Rescorer doesn't support n-best lists as an input yet. You need to provide two files with parallel sentences in plain text format with `-t` option. You can find a use case for `marian-scorer` in our regression tests: https://github.com/marian-nmt/marian-regression-tests/blob/master/tests/scorer/scores/test_scores.sh

**Comment by emjotde** [comment]
Supporting n-best lists is planned, that's why this issue is open. It will be closed once we add this functionality.

**Comment by lkfo415579** [comment]
@snukky  thanks, do you use one single model to do the **reranking**(have you ever tried L2R-R2L or source->target-target-source reranking method)? what is the mechanism or algorithm behind the **marian-rescorer**?

**Comment by snukky** [comment]
By default, **marian-scorer** returns a total sentence-level cost in the form of the cross entropy between the source and target sentences (see `--summary` for more options). The mechanism behind is equivalent to what is performed during validation.

I use the rescorer for a simple reranking with one or multiple models. If I want to use multiple models, I rescore with each model separately.

**Comment by emjotde** [comment]
BTW, the rescorer is multi-gpu now, as usual with `--devices 0 1 2 3`

**Comment by emjotde** [comment]
Works now last argument for `-t a b` is an n-best list and `--n-best` is given. It is also possible to define the feature name like `--n-best-feature FN0`.
